Fix 25798 : Allow "partial/prefix" filter for health_status, exec_create and exec_start events

vdemeester opened the pull request:
Fixes #25798, by using partial filtering for `health_status`, `exec_create` and `exec_start` events if there is no `:` in the filter 🐅.

<del>When we filter event, we do an exact match. Thus if the Action content is dynamic, we are filtering out some events. The Attributes field is there to pass additional data with events.</del>


<del>Fixing exec_create, exec_start and health_status event in that matter. A possible follow-up of this would be to force a type for Action in these methods to make sure we pass predefined actions.</del>


**\- How to verify it**

In one terminal:

``` bash
docker events --filter event=health_status
```

In another terminal;

``` bash
docker run --name=test -d --health-cmd='stat /etc/passwd || exit 1' --health-interval=2s   busybox sleep 1d
```

**\- Expected**

``` bash
$ docker events --filter event=health_status

2016-08-17T12:10:34.337301145Z container health_status: healthy da1bab7bc91a424cd8e0571808c1b4638a903a8508eb6612e9cdfc49f9dde555 (image=busybox, name=test)


$ docker events --filter event=exec_create

2016-08-17T12:10:34.198347963Z container exec_create: /bin/sh -c stat /etc/passwd || exit 1 da1bab7bc91a424cd8e0571808c1b4638a903a8508eb6612e9cdfc49f9dde555 (image=busybox, name=test)
2016-08-17T12:10:36.338156681Z container exec_create: /bin/sh -c stat /etc/passwd || exit 1 da1bab7bc91a424cd8e0571808c1b4638a903a8508eb6612e9cdfc49f9dde555 (image=busybox, name=test)


$ docker events --filter event=exec_start

2016-08-17T12:10:34.198549073Z container exec_start: /bin/sh -c stat /etc/passwd || exit 1 da1bab7bc91a424cd8e0571808c1b4638a903a8508eb6612e9cdfc49f9dde555 (image=busybox, name=test)
2016-08-17T12:10:36.338214034Z container exec_start: /bin/sh -c stat /etc/passwd || exit 1 da1bab7bc91a424cd8e0571808c1b4638a903a8508eb6612e9cdfc49f9dde555 (image=busybox, name=test)
```

**\- Todo**
- [x] Add an integrations test for that.
- [x] Decide if the map key are the right one for these cases.

/cc @thaJeztah @icecrime @tiborvass @stevvooe @calavera @cpuguy83 

🐸

Signed-off-by: Vincent Demeester vincent@sbr.pm

thaJeztah commented:
"status" and "command" sound good to me

jhorwit2 commented:
Thoughts on changing `status` since it's a [deprecated message field](https://github.com/docker/engine-api/blob/master/types/events/events.go#L32)? This PR would mean we have a deprecated `status` and a non-deprecated `status` for `container` events.

vdemeester commented:
@jhorwit2 good point :sweat_smile:

stevvooe commented on daemon/health.go:
```diff
@@ -136,7 +138,9 @@ func handleProbeResult(d *Daemon, c *container.Container, result *types.Healthch
 
 	if oldStatus != h.Status {
-		d.LogContainerEvent(c, "health_status: "+h.Status)
+		d.LogContainerEventWithAttributes(c, "health_status", map[string]string{
```
This is an API change. We'll have to update consumers in swarmkit that take health events.

vdemeester replied:
ah true :+1:

vdemeester commented:
Putting the `needs-attention` flag 👼

thaJeztah commented:
Design LGTM, but ping @stevvooe PTAL for the changes required in SwarmKit

thaJeztah commented:
oh, actually, do we need to do something w.r.t. https://github.com/docker/docker/pull/25800#issuecomment-240417315 ?

thaJeztah commented:
ping @talex5 could you have a look as well?

talex5 commented:
@thaJeztah: LGTM

thaJeztah commented:
This is the issue we need to think about: https://github.com/docker/docker/blob/master/docs/deprecated.md#ambiguous-event-fields-in-api

Anyone have a better name?

thaJeztah commented:
We discussed this in the maintainers review, and we're ok with this PR BUT we need a different name 😞

vdemeester commented:
@thaJeztah tried out "state" instead of status 👼 (the *state* of the *health_status*, make probably a little more sence than the *status* of the *health_status*)

tianon commented:
Do you think anyone's currently doing something like `docker events --filter 'event=health_status: healthy'` to where we might need some compatibility code to handle that (especially since that's the only existing way to consume these events easily)?

thaJeztah commented:
Possibly this should be versioned in the API?

We discussed this, and we do want this change as it's really a bug fix

thaJeztah commented:
We're ok with naming this "state" btw

thaJeztah commented:
Discussing this further; to be really safe we should;

- version this change in the API
- implement a "match" in the client to allow backward compatibility

vdemeester commented:
Sounds good to me 😇
@tianon : we kinda do it on traefik so I guess several project does so I do agree we should have some backward compatibility thingy

vdemeester commented on api/server/router/system/system_routes.go:
```diff
@@ -143,6 +145,40 @@ func (s *systemRouter) getEvents(ctx context.Context, w http.ResponseWriter, r *
+		values := strings.SplitN(event, ":", 2)
+		if len(values) == 2 {
+			value := strings.TrimSpace(values[1])
+			ef.Add("label", fmt.Sprintf("%s=%s", attribute, value))
```
This is a little bit hacky as `labels` are, in the events, store at the same level as `attributes`, so we can filter on any additionnal attributes using the label filter.

vdemeester reviewed (commented):
Updated the PR to add _versionned_ filters, using a little hack :stuck_out_tongue_closed_eyes:

vdemeester commented:
Also one question *arise*, is the way to only filter `health_status` event that are `healthy` for example. It is possible to do so, using the hack I'm using : `--filter event="health_status: healthy"` would now be `--filter event=health_status --filter label=healthy` — but it feels really weird to be able to do so with labels (and it's undocumented too).

vdemeester commented:
Let's redo that in a less *intrusive* way. If `health_status` is provided as an event filter (without a `:`), then we'll filter `health_status: …` event using *prefix* instead of full match. It does not require any changes of the API and let's us take time to refactor some part of the events 👼

vdemeester commented:
Updated to not change the API, or hacking around at all.
/cc @thaJeztah @tiborvass @cpuguy83 👼

thaJeztah commented on daemon/events/filter.go:
```diff
@@ -29,6 +29,24 @@ func (ef *Filter) Include(ev events.Message) bool {
+	// #25798 if an event filter contains either health_status, exec_create or exec_start without a colon
+	// Let's to a FuzzyMatch instead of an ExactMatch.
+	if ef.filterContains("event", map[string]struct{}{"health_status": {}, "exec_create": {}, "exec_start": {}}) {
+		return ef.filter.FuzzyMatch("event", ev.Action)
```
We should really rename `FuzzyMatch` to `PrefixMatch` as it's not doing anything "fuzzy", only optimizing a prefix match

vdemeester replied:
That's true :+1:

thaJeztah reviewed (changes requested):
changes LGTM, but a test would be nice to have

vdemeester committed:
Fix some event filtering

Make it possible to use health_status, exec_start and exec_create as
is in event filter. This way, using `health_status` as filter will allow
to get all health_status events (healthy, unhealthy, …) instead of
having to us all combination (`health_status: healthy`, `health_status:
unhealthy`, …). Same goes for `exec_start` and `exec_create`.

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

thaJeztah reviewed (approved):
LGTM

cpuguy83 commented:
LGTM

mlilback commented:
vdemeester: what is meant by "It does not require any changes of the API and let's us take time to refactor some part of the events"?

I'm running into this problem when observing exec_start via the API. Was an issue created for this refactoring?
